### ISSUE TITLE
Postfix - SASL security options - Incorrect pre-filled value

### ISSUE THREAD

**Issue by shoulders** [webmin/webmin#2146]
|  SYSTEM INFORMATION||
|----------------------|---------------------------|
|  OS type and version |     Ubuntu Linux 22.04.4     |
|  Webmin version      |     2.111      |
|  Usermin version     |     2.010     |
|  Virtualmin version  |     7.10.0       |
|  Theme version       |     21.10      |
|  Package updates     |      3  package updates are available   |

### background

while setting up my Postfix server I cam along this issue. I have not played around with postfix settings before and I checked this on my live site which I definately have not touched.

**In the GUI**

`Webmin --> Postfix Mail Server --> SMTP Client Options --> SASL security options: noplaintext, noanonymous`

![image](https://github.com/webmin/webmin/assets/319997/54b37f53-1f2b-4f67-b461-12424911c2e1)

- there are 2 values and they are separated by a comma

**in /etc/postfix/main.cf`**

The corresponding value is as follows `smtpd_sasl_security_options = noanonymous`

![image](https://github.com/webmin/webmin/assets/319997/9a4db1b3-47de-4508-b77e-837ab3fa6341)

- there is only one value

### the issue

- There is an inconsistency between what is in the config and the GUI
- When I save the the page (as is with noplaintext, noanonymous) `Webmin --> Postfix Mail Server --> SMTP Client Options` this wipes the  `smtpd_sasl_security_options` from the config

### Cause

- I don't know why the 2 locations are different (GUI and config file)
- but I do know the presense of the comma in the GUI version is what causes the value to get wiped.

### Additional

- If I remove the comma and use `SASL security options: noplaintext noanonymous` the value saves as expected.
- When I tried to access the tooltip I got the following
   ![image](https://github.com/webmin/webmin/assets/319997/300551d1-bf2b-46f7-9d1e-00e59241329d)

**Comment by iliajie** [comment]
Hello,

`smtp_sasl_security_options` and `smtpd_sasl_security_options` aren't the same options.

They also have different defaults, i.e.:

```
smtp_sasl_security_options (default: noplaintext, noanonymous)
```

vs

```
smtpd_sasl_security_options (default: noanonymous)
```

**Comment by shoulders** [comment]
let me recheck i pasted the right values in. I will post back shortly

**Comment by iliajie** [comment]
Nevertheless, missing help file was a bug. It is fixed now.

Thanks for the heads up!

**Comment by shoulders** [comment]
@iliajie You are right they are different settings but I think there is still a bug. I have tried to re-outline it below now I know those 2 values have a very similar name.

### New test on my dev server

I checked the **SASL security options** field for reference
 html for the field `name="smtp_sasl_security_options_def"`

**I added chicken**

![image](https://github.com/webmin/webmin/assets/319997/f3d83038-1029-44fe-9dac-ff24a14cc468)

--> 

![image](https://github.com/webmin/webmin/assets/319997/0f136fd1-31ca-4273-9fdd-70773b46684f)

- So this seems to correctly work for  `smtp_sasl_security_options` 
- `smtpd_sasl_security_options` does not exist, I cannot find the option in the GUI for this to clarify how this was removed but maybe this was something I did.


### New site (default config file)

1. `smtp_sasl_security_options` does not exist in the default config after virtualmin but has entries invalid in the GUI
  ![image](https://github.com/webmin/webmin/assets/319997/f09a0730-79bf-451c-8077-5cd3809562f5)
2.  in the config /etc/postfic/main.cf there is no setting with `noplaintext, noanonymous`
3. `smtpd_sasl_security_options` = noanonymous
   - this is present in the default config
   - I do not know where this setting is in the GUI

### New Conclusions
- the value in the GUI for `smtp_sasl_security_options` (noplaintext, noanonymous)
  - does not exist in the default /etc/postfic/main.cf after virtualmin installation
  - has a comma in it so is invalid
  - This might mean it was not applied by virtualmin correctly during the setup phase because there was comma in the values.
  - `noplaintext, noanonymous` is getting pulled from somewhere and it is not the config file until it has been applied via `SMTP Client options` page.
- `smtpd_sasl_security_options` = noanonymous
   - this is present in the default config file
   - I dont know how this value was removed because I don't know where the corresponding one, if any in the GUI, so this might of been me.
- the tooltip = glad it is fixed

**Comment by iliajie** [comment]
> * `smtp_sasl_security_options` does not exist in the default config after virtualmin but has entries invalid in the GUI
>   ![image](https://private-user-images.githubusercontent.com/319997/328024605-f09a0730-79bf-451c-8077-5cd3809562f5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQ5MjMxMDIsIm5iZiI6MTcxNDkyMjgwMiwicGF0aCI6Ii8zMTk5OTcvMzI4MDI0NjA1LWYwOWEwNzMwLTc5YmYtNDUxYy04MDc3LTVjZDM4MDk1NjJmNS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNTA1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDUwNVQxNTI2NDJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01ZDFkZjJjZjRjZWJlN2QxMDExOTg4ZTdkY2VmNDI4MTVjNWVlNzRiYzVjYWMxZTMwMGFiMGZhM2U4ZTQ4ZmQ1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.7BZU41j9Jkgtl-ux210JiS-ALUPzB4vp7Fi_Ye5yInw)
> * in the config /etc/postfic/main.cf there is no setting with `noplaintext, noanonymous`

It looks good to me. Even though it isn't represented on the config, `noplaintext` and `noanonymous` are defaults for `smtp_sasl_security_options` option.



> `smtpd_sasl_security_options` = noanonymous
> * this is present in the default config
> * I do not know where this setting is in the GUI

It's in **Servers ⇾ Postfix Mail Server: SMTP Authentication And Encryption** page, under **SMTP security options**.

**Comment by iliajie** [comment]
> * has a comma in it so is invalid

Not according to the [docs](https://www.postfix.org/SASL_README.html).

**Comment by iliajie** [comment]
> * `noplaintext, noanonymous` is getting pulled from somewhere and it is not the config file until it has been applied via `SMTP Client options` page.

```
postconf -d | grep smtp_sasl_security_options
```

**Comment by iliajie** [comment]
> I checked the **SASL security options** field for reference html for the field `name="smtp_sasl_security_options_def"`
> 
> **I added chicken**
> 
> ![image](https://private-user-images.githubusercontent.com/319997/328023785-f3d83038-1029-44fe-9dac-ff24a14cc468.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQ5MjMxMDIsIm5iZiI6MTcxNDkyMjgwMiwicGF0aCI6Ii8zMTk5OTcvMzI4MDIzNzg1LWYzZDgzMDM4LTEwMjktNDRmZS05ZGFjLWZmMjRhMTRjYzQ2OC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNTA1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDUwNVQxNTI2NDJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iMTVkN2FiNzMyNzRlNzM2YmU3YThiMzNkOGNlNmVkYjJlODYzNmJkNWY4YjdhNjJiOGNmODk5YmJhN2NjMTBkJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.vg4YEFglENdp5BOngo5lC6K24g-Ad5n3-CjT15BjVyg)

@jcameron Jamie, why adding "chicken" is possible?

**Comment by shoulders** [comment]
## Comma issue
Webmin --> Postfix Mail Server --> SMTP Client Options --> SASL security options

![image](https://github.com/webmin/webmin/assets/319997/ab565060-3884-4b32-9f22-31661d223b85)
-->
![image](https://github.com/webmin/webmin/assets/319997/e42b96af-61b8-4ef4-8f71-777f67b6faaf)
`smtp_sasl_security_options` = normal

![image](https://github.com/webmin/webmin/assets/319997/e09af1fa-61c7-4710-942d-eb78ac139676)
-->
![image](https://github.com/webmin/webmin/assets/319997/a1d172a3-7682-4b9c-9461-1e3a11b7dc13)
`smtp_sasl_security_options` = missing

I am sure there were other boxes where it did not like the comman.

**Comment by iliajie** [comment]
I just don't understand your last comment. I don't see any bugs there!

Please check the docs:

https://www.postfix.org/SASL_README.html

**Comment by shoulders** [comment]
- when i save the first one, `smtp_sasl_security_options = noplaintext noanonymous` gets added to the `/etc/postfix/main.cf` file
- when i save the second one with the comma in it, it does not get saved to the  `/etc/postfix/main.cf`

**So I am guessing now that the following is true**
- the SMTP Client Options is pulling the value from `postconf -d` and not directly from the config file
- because `noplaintext, noanonymous` is exactly the same as the default it does not get written to the config.

= not a bug, just did not expect this behaviour. I assumed all configs were pulled from the text configs files not from the live service.

If the default value was being used I would of expected this.
![image](https://github.com/webmin/webmin/assets/319997/3118234f-7cf7-4ff0-b5a1-5c29fd2e1b87)

**smtpd_sasl_security_options**

- I might be stupid, but I cannot find this option in `SMTP Client Options`
- I inspected the html and search for it just incase I missed it.
- I also inspected SMTP Server Options

@iliajie thanks for all of your time

**Comment by iliajie** [comment]
> * when i save the first one, `smtp_sasl_security_options = noplaintext noanonymous` gets added to the `/etc/postfix/main.cf` file
> * when i save the second one with the comma in it, it does not get saved to the  `/etc/postfix/main.cf`

I cannot reproduce it. It worked for me on both Rocky 9.3 and Ubuntu 24.04.

> * I cannot find this option in `SMTP Client Options`

The option name for `smtpd_sasl_security_options` is called **SMTP security options**.

<img width="1427" alt="image" src="https://github.com/webmin/webmin/assets/4426533/29d6396e-2ced-42c1-90c5-bfcc506be381">

**Comment by shoulders** [comment]
thanks for the option, I would never of found them because these are split rather than a freetype input box :smile

the issue with the comma is **NOT** a bug, but changeing the field to this format would stop people like me doing daft stuff:
![image](https://github.com/webmin/webmin/assets/319997/3118234f-7cf7-4ff0-b5a1-5c29fd2e1b87)

**But just one last roll of the dice to help clear things up**

- `noplaintext,noanonymous` shows in the config. This has only a comma.
- `noplaintext noanonymous` shows in the config. This only has a space.
- `noplaintext, noanonymous` does not show in config. This has comma followed by a space. This exactly matches the 
- "" (i,.e empty string) shows in config as follows
    ![image](https://github.com/webmin/webmin/assets/319997/51ba540d-4239-4a5a-b196-49ce3a87ccaa)

If I now manually wipe this line and reload the postfix configuration, `noplaintext, noanonymous` is restored.
or
if I apply the value via the GUI `noplaintext, noanonymous` the line in the config shown above will be removed as this also returns back to the default settings.

`postconf -d | grep smtp_sasl_security_options`

![image](https://github.com/webmin/webmin/assets/319997/2504ca13-f0d3-4613-a7ae-fac864c729d4)

**Comment by iliajie** [comment]
> but changeing the field to this format would stop people like me doing daft stuff:

Do you suggest changing to use spaces on Webmin side, even though documentation and Postfix defaults use commas?

**Comment by iliajie** [comment]
> If I now manually wipe this line and reload the postfix configuration, `noplaintext, noanonymous` is restored.

What exact problem are we solving? Couldn't Postfix interpret correctly both commas and spaces as params separator?

**Comment by shoulders** [comment]
>Do you suggest changing to use spaces on Webmin side, even though documentation and Postfix defaults use commas?

No. The documentation specifies spaces and commas are acceptable. I am not sure there is a preference but it is not an issue here as both work.

> Couldn't Postfix interpret correctly both commas and spaces as params separator?

It can and does, this is not an issue.

### Key points

- postfix is fine and working as it should. it accepts spaces and commas as value separators
- earlier I made incorrect assumption that commas in the values when submitted through webmin were causing an issue. this is not true.
- when you enter a custom value in the GUI you expect it to be present in the relevant config file unless you have selected the default option which means it might or might not be present as defaults sometimes are not present in the config as it is not required.
- if you have an empty string in the config file you don't expect the default values for this option to be in postconf -d. (However this is a postfix thing and I am a windows user)

### the issue

- When you input the value `noplaintext, noanonymous` into  **SASL security options (smtp_sasl_security_options)** the value will stay present in the GUI but will be removed from the `/etc/postfix/main.cf`
- This is because it is EXACTLY the same as the default value including the single comma followed by a space. Any other variation of this value pair or any other values will result in the value being written to the config file.
- If you do a `save and apply` from `SMTP Client Options` with a blank `smtp_sasl_security_options` the following line will be created in the config
   `smtp_sasl_security_options = `
  - This still allows the **postconf -d** values to be `noplaintext, noanonymous`  which is the system default so would not cause any functional issues.

### So the following logic is true for `smtp_sasl_security_options`

**Fresh Virtualmin install**
> GUI: noplaintext, noanonymous
config file: nothing
postconf -d: noplaintext, noanonymous

**SMTP Client Options - after initial save and apply but with no changes**
> GUI: noplaintext, noanonymous
config file: nothing
postconf -d: noplaintext, noanonymous

**SMTP Client Options - removed values (closest to restoring defaults) - after save and apply**
> GUI: noplaintext, noanonymous
config file: smtp_sasl_security_options = 
postconf -d: noplaintext, noanonymous

So as you can see, if you use the value `noplaintext, noanonymous` your input will never appear in the config file even though you have specified them in the in the GUI which is confusing especially since there is no `default option` so you assume that any values you enter will always appear in the config.

I appreciate this is an edge case but it got me because while researching my settings I was recommend to put `noplaintext, noanonymous` in for the best configuration but it is now the default.

You cannot return to a truly default state without manually deleting the `smtp_sasl_security_options = ` from the config file or manually entering `smtp_sasl_security_options` in the GUI. Both of which involve the config line getting deleted.

### proposed solution

change the input field style to the following which has acknowledgement of the default state which many other options have.

![image](https://github.com/webmin/webmin/assets/319997/7d04ecbd-1a2d-4382-b70b-052b41a03eab)

- if there is no value specified in the config file then this will also trigger the default radio to be selected.
- These assume an exact pattern match of `noplaintext, noanonymous`
  - if the exact default value is entered manually in to the GUI then this will also trigger the default radio to be selected.
  - if the exact default value is entered manually in to the config then this will also trigger the default radio to be selected.
- if the default values are entered in the GUI but are of a slight variance, these will appear as user custom inputted value.

p.s. I hope this helps explain this edge case and why i cause me a few issues.

**Comment by iliajie** [comment]
> * if you have an empty string in the config file you don't expect the default values for this option to be in postconf -d

What is the difference between the `postconf` and `postconf -d` commands? What does the `-d` flag do?

**Comment by shoulders** [comment]
> What is the difference between the postconf and postconf -d commands? What does the -d flag do?

nothing. Where I refer to **postconf / postconf -d** it means the same thing, it will of been a typo. It was a way of stating the actual values of the server rather than those just of the config file. The live server will have all of the default settings for thos options where there was no setting in the config file.

I took the command from one you posted earlier

```
postconf -d
postconf -d | grep smtp_sasl_security_options
```

![328035856-2504ca13-f0d3-4613-a7ae-fac864c729d4](https://github.com/webmin/webmin/assets/319997/943871a9-825c-4e02-981e-351ce122dec3)

**Comment by iliajie** [comment]
> nothing. Where I refer to **postconf / postconf -d** it means the same thing, it will of been a typo.

These commands' outputs are fundamentally different.

That said, the absence of `smtp_sasl_security_options` in the config and having `smtp_sasl_security_options` set to empty string do not have the same effect .. I think, right? 🙂

**Comment by shoulders** [comment]
> These commands' outputs are fundamentally different.

fair enough

> That said, the absence of smtp_sasl_security_options in the config and having smtp_sasl_security_options set to empty string do not have the same effect .. I think, right? 

Correct, they both cause the default value of 'noplaintext, noanonymous' to be used

**Comment by iliajie** [comment]
> they both cause the default value of `noplaintext, noanonymous` to be used

If you set `smtp_sasl_security_options = ` (i.e. empty string) in `main.cf` what is the output of:


```
postconf | grep smtp_sasl_security_options
```

vs

```
postconf -d | grep smtp_sasl_security_options
```

**Comment by shoulders** [comment]
- I set the string to 
  ![image](https://github.com/webmin/webmin/assets/319997/bd75fd36-3176-4a22-a689-35beeb49aad4)
- reloaded the postfix configuration

> postconf | grep smtp_sasl_security_options

![image](https://github.com/webmin/webmin/assets/319997/23b0b63b-fce5-45d8-a2af-dd32e8157712)

> postconf -d | grep smtp_sasl_security_options

![image](https://github.com/webmin/webmin/assets/319997/3eb66176-9fa5-47ef-8682-e99b5e944243)


**p.s.** I see the absolute difference in commands, I always used `postfix -d` and I will do some Linux studying soon 😄

**Comment by iliaross** [comment]
Reopened, as this was closed by an automated cleanup run. @shoulders, sorry for the noise, and thanks for the contribution. We definitely didn't mean any disrespect.

If it turns out the issue is already resolved for you, you're welcome to close it.
